### ISSUE TITLE
FreeRDP disconnects in the middle of session on Server 2008 R2 only when using gateway

### ISSUE THREAD

**Issue by uhtechteam** [FreeRDP/FreeRDP#5048]
I have found that xfreerdp crashes in the middle of my session (more quickly when it is inactive even for less than the Windows session timeout, less quickly when I'm actively using it, and seemingly not at all when it's minimized) when I connect to a 2008 R2 server **ONLY via a gateway**.  The version I have is a recent pull, **2.0.0-dev4 (b216520d5)** but I have seen this with a few previous versions as well.  When I don't use the gateway, it does not crash.  Windows clients connecting through the gateway don't have this issue.

My command to connect is:
`xfreerdp /v:myserver /cert-ignore /f /d:'mydomain' /g:'mygateway'`

Log of connect and disconnect is attached.
[freerdpcrash3.txt](https://github.com/FreeRDP/FreeRDP/files/2605116/freerdpcrash3.txt)

**Comment by akallabeth** [comment]
There is a gateway related pull awaiting review, might already fix your issue. Can you try that out?

**Comment by uhtechteam** [comment]
@akallabeth is it in the master branch? As you may recall from the other issue I'm on ThinStation and getting a specific git pull in there is really complicated apparently.

**Comment by akallabeth** [comment]
@uhtechteam you can try master, we've merged a lot for the `2.0.0-rc4` version, but the one I'm referring to is in #5022

**Comment by uhtechteam** [comment]
Pulling latest version today gave me 2.0.0-dev5.  Is that in there?

**Comment by uhtechteam** [comment]
@akallabeth Today's 2.0.0-dev5 didn't fix :(

**Comment by akallabeth** [comment]
@uhtechteam And did it crash or disconnect? Please clarify, the messages in the log look like some disconnection issue, but due to the missing error messages I can't be sure.

**Comment by uhtechteam** [comment]
I used that redirect string but without the “tee”. It disconnected and showed the same connection error.  I’ll generate a log again.

**Comment by akallabeth** [comment]
@uhtechteam the referenced pull #5022 has been merged, so you may want to try with that included.

**Comment by uhtechteam** [comment]
@akallabeth It still disconnects and the 2>&1 doesn't work on ThinStation for some odd reason.  It makes a zero length file and then just sits and does nothing.

**Comment by akallabeth** [comment]
@uhtechteam looks like there was a flaw in the last rdg pull that will be fixed with #5064 
Hope to get this merged soon.

**Comment by uhtechteam** [comment]
Yeah, I pulled 2.0.0-dev5 (8ae6b463e) and now I get “Error in GnuTSL initialization: Failed to acquire random data” and it won’t connect at all (Error: protocol security negotiation or connection failure). The version from 2 days ago was better.  I’ll wait for the next merge.

**Comment by uhtechteam** [comment]
Yeah LOL, sorry I was retyping from my other screen.  Here's the whole error:

Error in GnuTLS initialization: Failed to acquire random data.
[15:50:10:146] [10404:10405] [INFO][com.freerdp.client.common.cmdline] - loading channelEx cliprdr
[15:50:10:230] [10404:10405] [ERROR][com.freerdp.core.nego] - Protocol Security Negotiation Failure
[15:50:10:230] [10404:10405] [ERROR][com.freerdp.core] - freerdp_set_last_error ERRCONNECT_SECURITY_NEGO_CONNECT_FAILED [0x0002000C]
[15:50:10:230] [10404:10405] [ERROR][com.freerdp.core.connection] - Error: protocol security negotiation or connection failure

The GnuTLS is a new issue as of this pull, I think.

**Comment by akallabeth** [comment]
@uhtechteam still there are referenced a few things (e.g https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=759590;msg=30 which is a bit older but unresolved)
The TLS layer did not change in quite a while (except some changes for libressl and openssl 1.1)

**Comment by uhtechteam** [comment]
Do you think that's related to the whole gateway disconnect issue?

**Comment by akallabeth** [comment]
@uhtechteam might be. tested RDG and HTTP gateway functionality today after #5064 was merged.

**Comment by uhtechteam** [comment]
Aargh @akallabeth same issue and "2>&1 |tee log.txt" on my command line just makes a zero length file and xfreerdp doesn't run.  Back to the drawing board.

**Comment by akallabeth** [comment]
@uhtechteam I'd suggest you try to find out if it works on a normal linux installation or not. if not, we can at least get some useful logging, otherwise we know it is your setup.

**Comment by akallabeth** [comment]
Any update on this one?

**Comment by uhtechteam** [comment]
So far I’ve only been able to work with it on the ThinStation Linux.  The only other Linux I have around is Debian (Raspberry Pi/Raspbian for ARM and the Raspberry Pi Desktop version for Intel) and I can use xfreerdp on there after installing with apt-get, but I’m sure that version is much older than the current version on github.

Do you think a test of xfreerdp in that environment with my gateway would be at all informative?

**Comment by akallabeth** [comment]
Yes, you can compile our nightlies with dpkg-buildpackage on the raspberry https://github.com/FreeRDP/FreeRDP/wiki/Compilation
That will give good debug information

**Comment by uhtechteam** [comment]
@akallabeth I managed to get the "2>&1 |tee log.txt" working if I sudo tee... It actually is prompting me on the command line for a gateway username and password which I provide (blind) and then it does connect.  Silly me.
Do you want me to compile for debug before using that method to generate the log file?

**Comment by uhtechteam** [comment]
Compiled with debug and ran with log level trace.  Here's the file.
[freerdplog-debug-trace.txt](https://github.com/FreeRDP/FreeRDP/files/2650648/freerdplog-debug-trace.txt)

**Comment by akallabeth** [comment]
@uhtechteam Ok, of course the error occurs without additional log messages :/ I'll prepare a branch with additional log messages to trace that a bit further.
btw, if you need `sudo tee` you most likely wrote to the wrong folder ;) try something like `tee /tmp/log.txt` then this should work.

**Comment by akallabeth** [comment]
@uhtechteam ok, created https://github.com/akallabeth/FreeRDP/tree/gw_logging with some additional log messages for error conditions currently not writing any.

**Comment by uhtechteam** [comment]
@akallabeth I am trying to pull by using "git clone git://github.com/akallabeth/FreeRDP.git" and it's failing as follows:

```
prt-get: updating /ts/ports/opt/freerdp
=======> Building '/ts/ports/opt/freerdp/freerdp#2.0-dev-master.pkg.tar.xz'.
+ build
+ git clone git://github.com/akallabeth/FreeRDP.git
Cloning into 'FreeRDP'...
+ cd FreeRDP
+ cmake . -DCMAKE_INSTALL_PREFIX=/usr -DCMAKE_INSTALL_LIBDIR=lib -DDOCBOOKXSL_DIR=/usr/share/sgml/docbook/xsl-ns-stylesheets -DCMAKE_BUILD_TYPE=Debug -DWITH_SANITIZE_ADDRESS=ON -DWITH_PCSC=ON -DBUILD_SHARED_LIBS=ON -DWITH_CHANNELS=ON -DWITH_CHANNELS_PRINT=ON -DWITH_SERVER=OFF -DWITH_CLIENT=ON -DWITH_PULSE=ON -DWITH_ALSA=ON -DWITH_CUPS=ON -DWITH_FFMPEG=ON -DWITH_MBEDTLS=OFF -DWITH_OPENSLES=OFF -DCHANNEL_URBDRC_CLIENT=ON -DWITH_GSTREAMER_1_0=ON -DWITH_GSTREAMER_0_10=OFF -DWITH_DIRECTFB=OFF -DWITH_XINERAMA=ON -DWITH_CUNIT=OFF -DWITH_JPEG=ON -DWITH_IPP=OFF -DWITH_NPP=OFF -DWITH_MANPAGES=ON -DWITH_SAMPLE=ON -DWITH_x264=ON -DWITH_GSSAPI=ON -DWITH_FAAC=ON -DWITH_LIBSYSTEMD=OFF -DWITH_WAYLAND=OFF
-- The C compiler identification is GNU 5.3.0
-- The CXX compiler identification is GNU 5.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.28") 
-- FreeRDP GIT
-- Git Revision 743929909
-- Performing Test Wno-unused-result
-- Performing Test Wno-unused-result - Success
-- Performing Test Wno-unused-but-set-variable
-- Performing Test Wno-unused-but-set-variable - Success
-- Performing Test Wno-deprecated-declarations
-- Performing Test Wno-deprecated-declarations - Success
-- Performing Test Wno-deprecated-declarationsCXX
-- Performing Test Wno-deprecated-declarationsCXX - Success
-- Looking for include file fcntl.h
-- Looking for include file fcntl.h - found
-- Looking for include file unistd.h
-- Looking for include file unistd.h - found
-- Looking for include file execinfo.h
-- Looking for include file execinfo.h - found
-- Looking for include file stdint.h
-- Looking for include file stdint.h - found
-- Looking for include file inttypes.h
-- Looking for include file inttypes.h - found
-- Looking for include file sys/modem.h
-- Looking for include file sys/modem.h - not found
-- Looking for include file sys/filio.h
-- Looking for include file sys/filio.h - not found
-- Looking for include file sys/strtio.h
-- Looking for include file sys/strtio.h - not found
-- Looking for include file sys/select.h
-- Looking for include file sys/select.h - found
-- Performing Test HAVE_TM_GMTOFF
-- Performing Test HAVE_TM_GMTOFF - Success
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Looking for pthread_tryjoin_np in pthread
-- Looking for pthread_tryjoin_np in pthread - found
-- Looking for include file sys/eventfd.h
-- Looking for include file sys/eventfd.h - found
-- Looking for include file sys/eventfd.h
-- Looking for include file sys/eventfd.h - found
-- Looking for include file sys/timerfd.h
-- Looking for include file sys/timerfd.h - found
-- Looking for include file poll.h
-- Looking for include file poll.h - found
-- Finding recommended feature X11 for X11 (X11 client and server)
--     Disable feature X11 using "-DWITH_X11=OFF"
-- Found X11: /usr/lib/libX11.so  
-- Skipping recommended feature Wayland for Wayland (Wayland client)
-- Skipping optional feature DirectFB for DirectFB (DirectFB client)
--     Enable feature DirectFB using "-DWITH_DIRECTFB=ON"
-- Finding required feature ZLIB for compression (data compression)
-- Found ZLIB: /usr/lib/libz.so (found version "1.2.5") 
-- Finding required feature OpenSSL for cryptography (encryption, certificate validation, hashing functions)
-- Found OpenSSL: /usr/lib/libssl.so;/usr/lib/libcrypto.so (found version "1.1.0g") 
-- Skipping optional feature OpenSLES for multimedia (OpenSLES audio / video)
--     Enable feature OpenSLES using "-DWITH_OPENSLES=ON"
-- Finding recommended feature ALSA for sound (audio input, audio output and multimedia redirection)
--     Disable feature ALSA using "-DWITH_ALSA=OFF"
-- Found ALSA: /usr/lib/libasound.so (found version "1.1.5") 
-- Finding optional feature Pulse for sound (audio input, audio output and multimedia redirection)
-- Checking for module 'libpulse'
--   Found libpulse, version 11.1
-- Found Pulse: /usr/include  
-- Finding optional feature Cups for printing (printer device redirection)
-- Found Cups: /usr/lib/libcups.so (found version "2.2.6") 
-- Finding optional feature PCSC for smart card (smart card device redirection)
-- Found PCSC: /usr/lib/libpcsclite.so  
-- Finding recommended feature FFmpeg for multimedia (multimedia redirection, audio and video playback)
--     Disable feature FFmpeg using "-DWITH_FFMPEG=OFF"
-- Checking for module 'libavcodec'
--   Found libavcodec, version 57.24.102
-- Checking for module 'libavutil'
--   Found libavutil, version 55.17.103
-- Found FFmpeg: TRUE  
-- Skipping optional feature GStreamer_0_10 for multimedia (multimedia redirection, audio and video playback, gstreamer 0.10 version)
--     Enable feature GStreamer_0_10 using "-DWITH_GSTREAMER_0_10=ON"
-- Finding recommended feature GStreamer_1_0 for multimedia (multimedia redirection, audio and video playback)
--     Disable feature GStreamer_1_0 using "-DWITH_GSTREAMER_1_0=OFF"
-- Checking for module 'glib-2.0'
--   Found glib-2.0, version 2.54.3
-- Found Glib 
-- Checking for module 'gstreamer-1.0'
--   Found gstreamer-1.0, version 1.12.1
-- Checking for module 'gstreamer-base-1.0'
--   Found gstreamer-base-1.0, version 1.12.1
-- Checking for module 'gstreamer-app-1.0'
--   Found gstreamer-app-1.0, version 1.12.1
-- Checking for module 'gstreamer-audio-1.0'
--   Found gstreamer-audio-1.0, version 1.12.1
-- Checking for module 'gstreamer-fft-1.0'
--   Found gstreamer-fft-1.0, version 1.12.1
-- Checking for module 'gstreamer-pbutils-1.0'
--   Found gstreamer-pbutils-1.0, version 1.12.1
-- Checking for module 'gstreamer-video-1.0'
--   Found gstreamer-video-1.0, version 1.12.1
-- Found GSTREAMER_1_0: /usr/lib/libgstreamer-1.0.so  
-- Finding optional feature JPEG for codec (use JPEG library)
-- Found JPEG: /usr/lib/libjpeg.so  
-- Skipping optional feature OpenH264 for codec (use OpenH264 library)
--     Enable feature OpenH264 using "-DWITH_OPENH264=ON"
-- Skipping optional feature GSM for codec (GSM audio codec library)
--     Enable feature GSM using "-DWITH_GSM=ON"
-- Skipping optional feature IPP for performance (Intel Integrated Performance Primitives library)
--     Enable feature IPP using "-DWITH_IPP=ON"
-- FreeRDP GIT
-- Finding recommended feature XKBFile for X11 keyboard (X11 keyboard file extension)
--     Disable feature XKBFile using "-DWITH_XKBFILE=OFF"
-- Found XKBFile: /usr/lib/libxkbfile.so  
CMake Error at channels/client/CMakeLists.txt:27 (list):
  list sub-command REMOVE_DUPLICATES requires list to be present.


-- FreeRDP GIT
-- Found DocBookXSL: /usr/share/sgml/docbook/xsl-ns-stylesheets  
-- Finding required feature XShm for X11 shared memory (X11 shared memory extension)
-- Found XShm: /usr/lib/libXext.so  
-- Finding recommended feature Xinerama for multi-monitor (X11 multi-monitor extension)
--     Disable feature Xinerama using "-DWITH_XINERAMA=OFF"
-- Found Xinerama: /usr/lib/libXinerama.so  
-- Finding recommended feature Xext for X11 extension (X11 core extensions)
--     Disable feature Xext using "-DWITH_XEXT=OFF"
-- Found Xext: /usr/lib/libXext.so  
-- Finding recommended feature Xcursor for cursor (X11 cursor extension)
--     Disable feature Xcursor using "-DWITH_XCURSOR=OFF"
-- Found Xcursor: /usr/lib/libXcursor.so  
-- Finding recommended feature Xv for video (X11 video extension)
--     Disable feature Xv using "-DWITH_XV=OFF"
-- Found Xv: /usr/lib/libXv.so  
-- Finding recommended feature Xi for input (X11 input extension)
--     Disable feature Xi using "-DWITH_XI=OFF"
-- Found Xi: /usr/lib/libXi.so  
-- Looking for XITouchClass
-- Looking for XITouchClass - found
-- Finding recommended feature Xrender for rendering (X11 render extension)
--     Disable feature Xrender using "-DWITH_XRENDER=OFF"
-- Found Xrender: /usr/lib/libXrender.so  
-- Finding recommended feature Xfixes for X11 xfixes extension (Useful additions to the X11 core protocol)
--     Disable feature Xfixes using "-DWITH_XFIXES=OFF"
-- Found Xfixes: /usr/lib/libXfixes.so  
-- Configuring incomplete, errors occurred!
See also "/ts/ports/opt/freerdp/work/src/FreeRDP/CMakeFiles/CMakeOutput.log".
See also "/ts/ports/opt/freerdp/work/src/FreeRDP/CMakeFiles/CMakeError.log".
=======> ERROR: Building '/ts/ports/opt/freerdp/freerdp#2.0-dev-master.pkg.tar.xz' failed.

-- Packages where update failed
freerdp

```
Do I need to change flags or pull differently for this?

**Comment by uhtechteam** [comment]
@akallabeth I had to get some help to figure out how to grab this commit and voila...here's the updated log.
[new log with gw_logging.txt](https://github.com/FreeRDP/FreeRDP/files/2654207/new.log.with.gw_logging.txt)

**Comment by akallabeth** [comment]
@uhtechteam ok, that is now even stranger. You are properly connected, everything is working and suddenly the `SSL` connection is breaking up. (Can not read anything from the channel anymore)

**Comment by akallabeth** [comment]
@uhtechteam can you verify this setup actually works with any client (over a extended period of time), it is looking more and more like a network issue.

**Comment by uhtechteam** [comment]
What I can tell you for sure is that I get the same disconnect on any of these Linux clients running from anywhere, and that my Windows and Mac clients accessing the same terminal server and gateway don’t seem to disconnect at all.  It only happens in FreeRDP and only when going through the gateway.

Since it doesn’t seem to kick off the Windows and Mac clients, I don’t think it’s a network issue.

**Comment by akallabeth** [comment]
@uhtechteam which mac clients did you test with?

**Comment by uhtechteam** [comment]
We have a handful of people using the Microsoft RDP client on Mac.  Haven’t tried with FreeRDP.

**Comment by uhtechteam** [comment]
I also suspect that keep alive is somehow involved here, as if I am actively using it, it won’t disconnect as quickly, but if I walk away from my desk for a minute and a half, it will be disconnected by the time I get back.  That doesn’t happen with the MS client software.

**Comment by akallabeth** [comment]
@uhtechteam try with `/echo /network:auto` and check if that helps.

**Comment by uhtechteam** [comment]
@akallabeth one thing I can tell you for sure is that my fonts look a lot prettier with those switches :) not sure what they do exactly but testing at present.

**Comment by uhtechteam** [comment]
Looked pretty but disconnected.  Log attached but the only difference seems to be a new set of errors at the bottom:
[15:22:08:546] [3009:3012] [DEBUG][com.freerdp.core] - rdp_check_fds() - -1
[15:22:08:546] [3009:3012] [INFO][com.freerdp.client.common] - Network disconnect!
[15:22:08:551] [3009:3018] [DEBUG][com.freerdp.channels.rdpsnd.client] - Device already 
[large log for freerdp.zip](https://github.com/FreeRDP/FreeRDP/files/2657813/large.log.for.freerdp.zip)

**Comment by akallabeth** [comment]
@uhtechteam ok, interesting. we've merged the logger changes back to master in between. the error is still the same, some parsing error in a http response. Might need to dump that response to analyze.

**Comment by uhtechteam** [comment]
OK.  So should I go back to the regular master branch and then await your instructions for the next log type?

**Comment by uhtechteam** [comment]
Also, 1 potentially informative piece of information: this is a server farm we are using, and the blue bar at the top starts off showing the farm address (tsfarm...com) but after I minimize and maximize, it'll show the name of the actual server I'm connected to (currently ts1...com).  Not sure if that change or informational update is in any way related to the issue we are having but worth noting.

**Comment by uhtechteam** [comment]
@akallabeth just checking in to see if there is anything else I can do.

**Comment by akallabeth** [comment]
@uhtechteam still can't reproduce. no clue what is actually wrong (can't rule out configuration issues/software bugs/...)
The title update stuff is just informative, will have a look at it though (should be updated in place and not on next window hide/show)

**Comment by uhtechteam** [comment]
@akallabeth it's OK, in fullscreen that bar seems to have disappeared anyway which is fine with me.  Is there some kind of packet trace/dump I can do that will actually capture the traffic through the gateway that causes the disconnect?  That might be informative.

**Comment by akallabeth** [comment]
@uhtechteam complete setup description and gateway configuration might help. (Which os, all policy vales, ...)

**Comment by akallabeth** [comment]
Did you find any pattern in the disconnects? (Certain transferred data size, time, ...?)

**Comment by uhtechteam** [comment]
What I've identified is:

- It will eventually disconnect regardless
- When it is not being used (i.e. you walk away from the computer for 2 minutes) it will almost definitely have disconnected when you come back
- If you are using it actively and continuously, it will take longer to disconnect
- If it does not drop off that way and it goes to the Windows lock screen, it will sit there for a long time and not disconnect from that

It's a bizarre report but those have been my findings.

**Comment by akallabeth** [comment]
@uhtechteam Ok, finally managed to reproduce such a disconnect:
1. Watching youtube, don't do anything on the computer
1. The local image freezes
1. The session disconnects (after ~30 seconds to 1 minute)

What is interesting though is that this only happens with legacy gdi, as soon as `/rfx` or `/gfx` is uses I can no longer reproduce this.

**Comment by akallabeth** [comment]
@uhtechteam could you try if using `+heartbeat` makes any difference?

**Comment by akallabeth** [comment]
@uhtechteam ok, additional information: Can only reproduce with the `RPC` gateway implementation, the `HTTP` version is not affected.
(side note: the same server with the same gateway is running much smoother with `/gt:http` while `/gt:rpc` is a slide show)

**Comment by akallabeth** [comment]
@uhtechteam could you also check the value (and if it exists) of `HKLM\Software\Microsoft\Rpc\DefaultChannelLifetime`?

**Comment by uhtechteam** [comment]
Wow @akallabeth that was quite a report.  Testing now with `+heartbeat` and so far so good but I don't think I've been connected for long enough to know for sure.  I've got to give it a few hours at least :)

The key you referenced doesn't exist in the registry on the gateway.

I'm not sure I can connect as you suggested by http as when I connect I see the following:
```

[14:53:54:521] [17444:19175] [INFO][com.freerdp.core.gateway.rdg] - RD Gateway does not support HTTP transport.
[14:53:55:335] [17444:19175] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success
[14:53:55:436] [17444:19175] [WARN][com.freerdp.core.nego] - Error: SSL_NOT_ALLOWED_BY_SERVER
[14:53:56:684] [17444:19175] [INFO][com.freerdp.core.gateway.rdg] - RD Gateway does not support HTTP transport.
[14:53:56:368] [17444:19175] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success
[14:53:56:468] [17444:19175] [WARN][com.freerdp.core.nego] - Error: SSL_NOT_ALLOWED_BY_SERVER
[14:53:57:701] [17444:19175] [INFO][com.freerdp.core.gateway.rdg] - RD Gateway does not support HTTP transport.
[14:53:57:388] [17444:19175] [INFO][com.freerdp.core.gateway.tsg] - TS Gateway Connection Success

```
Doesn't that mean HTTP isn't an option for me?

**Comment by uhtechteam** [comment]
@akallabeth it eventually booted me with `+heartbeat`.  Same disconnect message.

**Comment by uhtechteam** [comment]
Also dropped with /gfx for me.  Full command line was:
`xfreerdp /v:myserverfarm /cert-ignore /f /d:'mydomain' /g:'mygateway' /echo /network:auto +heartbeat /gfx`

**Comment by akallabeth** [comment]
@uhtechteam server2008 only supports `/rfx` (did the `/gfx` tests against a windows 10) but that needs to be enabled server side.
as for the http gateway stuff, yes, it is not activated for your configuration.

**Comment by uhtechteam** [comment]
Got it.  So the question is what should I ask our server people to change/enable?  Or should we concentrate on figuring out how to fix in the current configuration?

**Comment by akallabeth** [comment]
@uhtechteam you can enable rfx on the servers (not gateway) and add the /rfx flag client wise.

Gateway wise http would be nice, but I don't know when that was introduced (tested with a 2012 gw where I could reproduce)

**Comment by uhtechteam** [comment]
Working on that on our end.  In the meantime I guess the question is if it's worth tracking down & fixing the cause of the issue in an basic 2008 R2 configuration like mine or not.

**Comment by uhtechteam** [comment]
@akallabeth Here's what one of the techs from our MSP said:

1.  I don't really see a way to update the config on the S2008 gateway which would help us here.
There are limitations on what is supported here; and /rfx or RemoteFX is not really supported due to the VM being on a VMware host (this is more of a Hyper-V thing).

I'm looking into session timeouts and keep alive, to define them explicitly here, especially on the gateway.
For now, I didn't touch the keep alive setting, but we should look into that.

Can you see is there an option for xfreerdp (or a setting) where you could define the timeout (a version of keep alive); maybe something like "timeoutWarning"?

then

2.  I've explicitly defined this on the gateway (which it wasn't before):

![gw change screenshot](https://user-images.githubusercontent.com/44818626/50164459-67c16680-02a8-11e9-8190-78a87dadaf88.png)

But did not touch keep-alive, as it can respond differently to different types of clients, so want to be careful with it; it would be ideal if the TCs with issues can have this defined on their side (or just disabled, usually with value 0).

**Comment by akallabeth** [comment]
@uhtechteam the timeout is a server side thing that does not get communicated, so no, `xfreerdp` can not display such a warning.
You can create a event periodically sending mouse events though, that should keep the session `alive` if that is required.

**Comment by uhtechteam** [comment]
Interesting.  How would I do that?

**Comment by akallabeth** [comment]
Either externally with xdotool or similar (that just moves the mouse) or in code with a periodic timer just as in xf_disp.c

**Comment by akallabeth** [comment]
So, do I understand correctly that with these timeout settings the issues are solved?

**Comment by uhtechteam** [comment]
@akallabeth not really.  The server-side timeout settings changes don't prevent my connections from dropping.  I don't know how to implement the code you're referring to though I wonder if it would be best included in the RPC gateway code in xfreerdp as it would mitigate the disconnect issue for anyone using such a configuration.  It is still dropping off for me, but I added +auto-reconnect on my command line and that seems to mitigate the disconnects somewhat by automatically reconnecting quickly enough that it is not a major interruption.  I guess that's a half-workaround, not a fix, really.

**Comment by monkeytreehosting** [comment]
Hi,

I know this is a reasonably old thread, however I'm seeing what I believe to be this exact error; this seems to be very easy to replicate, however I can only find a couple of others having reported it online.

Setup:

OS: Ubuntu 18.04.2 LTS 
Connection: Remmina via official PPA

OS: NoTouch Desktop (tested on both k413 on HP T520 Thin Client and k419 on Rpi3B+)
Connection: FreeRDP version 2.0 (6be3ba4df7)

Connecting to:

- 2016 Windows TS cluster, via 2012R2 Gateway Cluster

I have only seen this disconnecting scenario since testing these linux based clients, both of which use FreeRDP. The below is from the error shown when the session dies in NoTouch Desktop using FreeRDP:

[ERROR][com.freerdp.core.transport] - BIO_read returned a system error 104: Connection reset by peer.
[ERROR][com.freerdp.core] - freerdp_check_fds() failed - 0
[ERROR][com.freerdp.client.x11] - Failed to check FreeRDP file descriptor
[INFO][com.freerdp.client.common] - Network disconnect!
FreeRDP Exitcode: 131
ErrorLevel: 131

When Remmina has issues, it reattempts connection and comes back in, however this happens about 5-10 times a day; sometimes the session just terminates.

To cause the disconnection, and thus replicate, I initiate a connection and then leave it with no input for a period of time; I've seen rare disconnects within 2mins, but generally see the disconnect following either 10-12mins, or 20-25mins. If input is given to the session, the connection is kept active and I have not experienced disconnect while actively working, however some other test clients have reported disconnects while typing away.

To confirm, we've been connecting to this hosted platform for 4-years without issue by using either Windows or Mac clients without issue.

I have just found another user, using remmina build with FreeRDP, who is experiencing the same issue too: https://www.reddit.com/r/Remmina/comments/bb5qh5/remmina_closes_connection_randomly/

I have also searched through the GitHub issues and only found a couple of examples, all referencing 2008 gateways, however a common theme here seems to be the RDP gateway.

We're currently in the process of implementing RDGW 2019, however this will be about a month before complete and able to test.

Any help, or advice would be greatly appreciated!

Thanks,

David

**Comment by monkeytreehosting** [comment]
Created new issue as my issue is using different RDGW version #5344

**Comment by monkeytreehosting** [comment]
@uhtechteam Did you get any further with this issue?

I'm still experiencing the same in my thread #5344 however have mitigated it for now with the auto-reconnect flag. As you said, this hides the issue opposed to fixing it.

We've recently implemented 2019 Gateway, so I'll start testing on that shortly too

**Comment by uhtechteam** [comment]
No, same here.  It disconnects momentarily and transparently reconnects right after that.  Since it functions OK, I’m just leaving it alone for now

**Comment by monkeytreehosting** [comment]
@uhtechteam Thanks for confirming.

Have you done any updates windows side yet? or planning too?

We're currently using Gateway on 2012R2 and looking to update to Gateway on 2019.

Also exploring HAProxy as the perimeter; linux to linux might sit nicely...

**Comment by uhtechteam** [comment]
Nope, we’ve left the entire Windows environment alone for now.  That will probably change soon, though.
